File: src/algorithms/cpp/binary.cpp
// Binary Tree Node Definition and Traversals in C++
#include <iostream>
#include <vector>
#include <string>

// Simple emit macro to prevent empty trace errors in sandbox
#define EMIT(type, id) std::cout << "{\"type\": \"" << type << "\", \"nodeIds\": [\"" << id << "\"]}\n"

struct Node {
    std::string value;
    std::string id;
    Node* left;
    Node* right;

    Node(std::string val, std::string nid) : value(val), id(nid), left(nullptr), right(nullptr) {}
};

void inorder(Node* root, std::vector<std::string>& path) {
    if (!root) return;
    inorder(root->left, path);
    EMIT("VISIT", root->id);
    path.push_back(root->value);
    inorder(root->right, path);
}

void preorder(Node* root, std::vector<std::string>& path) {
    if (!root) return;
    EMIT("VISIT", root->id);
    path.push_back(root->value);
    preorder(root->left, path);
    preorder(root->right, path);
}

void postorder(Node* root, std::vector<std::string>& path) {
    if (!root) return;
    postorder(root->left, path);
    postorder(root->right, path);
    EMIT("VISIT", root->id);
    path.push_back(root->value);
}

int main() {
    Node* root = new Node("+", "node-root");
    root->left = new Node("A", "node-A");
    root->right = new Node("*", "node-mul");
    root->right->left = new Node("B", "node-B");
    root->right->right = new Node("C", "node-C");

    std::vector<std::string> path;
    preorder(root, path);
    inorder(root, path);
    postorder(root, path);
    
    return 0;
}

File: src/algorithms/python/binary.py
# Binary Tree Node Definition and Traversals in Python
# Implements basic hierarchical binary node structure and depth-first/level-order traversals.

import json
import time


def emit(event_type, **kwargs):
    event = {"type": event_type}
    event.update(kwargs)
    print(json.dumps(event))


class Node:
    def __init__(self, value, id):
        self.value = value
        self.id = id
        self.left = None
        self.right = None


def inorder_traversal(root):
    if root is None:
        return []

    left_res = inorder_traversal(root.left)

    emit("VISIT", nodeIds=[root.id])
    time.sleep(0.1)  # Simulate some work

    right_res = inorder_traversal(root.right)
    return left_res + [root.value] + right_res


def preorder_traversal(root):
    if root is None:
        return []

    emit("VISIT", nodeIds=[root.id])
    time.sleep(0.1)

    left_res = preorder_traversal(root.left)
    right_res = preorder_traversal(root.right)
    return [root.value] + left_res + right_res


def postorder_traversal(root):
    if root is None:
        return []

    left_res = postorder_traversal(root.left)
    right_res = postorder_traversal(root.right)

    emit("VISIT", nodeIds=[root.id])
    time.sleep(0.1)

    return left_res + right_res + [root.value]


# Create standard expression binary tree template
root = Node("+", "node-root")
root.left = Node("A", "node-A")
root.right = Node("*", "node-mul")
root.right.left = Node("B", "node-B")
root.right.right = Node("C", "node-C")

emit("SYSTEM_LOG", level="INFO", message="Starting Pre-order Traversal")
preorder_traversal(root)

emit("SYSTEM_LOG", level="INFO", message="Starting In-order Traversal")
inorder_traversal(root)

emit("SYSTEM_LOG", level="INFO", message="Starting Post-order Traversal")
postorder_traversal(root)
